feat(core): document service (un)publish multiple locales

jhoward1994 opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the tests
- Create or update the documentation at https://github.com/strapi/documentation
- Refer to the issue you are closing in the PR description: Fix #issue
- Specify if the PR is ready to be merged or work in progress (by opening a draft PR)

Please ensure you read the Contributing Guide: https://github.com/strapi/strapi/blob/main/CONTRIBUTING.md
-->

### What does it do?

Allows arrays of locales to be passed to publish and unpublish in the document service

### Why is it needed?

So we can (un)publish multiple locales at once from the CM edit view

### How to test it?

Added API tests

### Related issue(s)/PR(s)

CONTENT-1903
DX-1335

Marc-Roig commented on packages/core/core/src/services/document-service/internationalization.ts:
```diff
@@ -37,26 +37,15 @@ const localeToLookup: Transform = (contentType, params) => {
     }
 
-    return assoc(['lookup', 'locale'], params.locale, params);
+    return assoc(['lookup', 'locale', '$in'], params.locale, params);
```
how does this work with relations?
I recall we used the document locale sometimes to transform relations, and I wonder if this might add some conflicts

jhoward1994 replied:
Do you mean specifically with a bulk locale publish (e.g. publish something that has a relation with another non i18n CT)?

Or do you mean for the other document service methods?
	-In which case, would making the change to only allows arrays in the params of (un)publish remove this risk?

Marc-Roig replied:
other documetn service methods, with the current typing, there might be users doing something like
`strapi.documents(uid).findOne(docId, { locale: ['en', 'fr']})`

It would not be breaking to allow that, but I don't think such api makes sense atm, so I would say for now I would only allow it on the (un)publish & discardDraft methods

Marc-Roig replied:
it can be as simple as throwing an error if passing something that is not a string on the other methods

alexandrebodin commented on packages/core/core/src/services/document-service/internationalization.ts:
```diff
@@ -37,26 +37,15 @@ const localeToLookup: Transform = (contentType, params) => {
     }
 
-    return assoc(['lookup', 'locale'], params.locale, params);
+    return assoc(['lookup', 'locale', '$in'], params.locale, params);
```
this might not be an array if it's not a string it might be sth else do we validate the param somewhere else ? 

the $in is not necessary by the way so I would simplify as the array in a where will be automatically managed

alexandrebodin commented:
In theory looks ok, (almost too simple :D) will required extensive QA for sure

Convly commented on packages/core/types/src/modules/documents/params/index.ts:
```diff
@@ -43,7 +43,7 @@ export type Pick<TSchemaUID extends UID.Schema, TKind extends Kind> = MatchAllIn
     [HasMember<TKind, 'status'>, PublicationStatus.Param],
     // Locale
-    [HasMember<TKind, 'locale'>, { locale?: Locale }], // TODO: also allow arrays ?
+    [HasMember<TKind, 'locale'>, { locale?: Locale | Locale[] }],
```
We could have followed the same approach as for the rest of the params with different notation (Locale.StringNotation, Locale.ArrayNotation and Locale.Any) so that methods can decide what format they accept. What do you think?

alexandrebodin commented on packages/core/core/src/services/document-service/internationalization.ts:
```diff
@@ -70,7 +55,7 @@ const localeToData: Transform = (contentType, params) => {
   }
 
-  if (params.locale) {
+  if (params.locale && typeof params.locale === 'string') {
```
what happens if locale is * here ?

jhoward1994 replied:
I think you're right and we shouldn't allow that. This is only used in the document service create
and update but we'd never want to put '*' into the data

Marc-Roig commented on packages/core/content-manager/admin/src/pages/ListView/components/BulkActions/PublishAction.tsx:
```diff
@@ -325,7 +325,8 @@ const SelectedEntriesModalContent = ({
       });
 
-      setPublishedCount(res.data.count);
+      // TODO fix for V5
+      setPublishedCount(res.data.data.count);
```
is the issue here that data is nested twice?

jhoward1994 replied:
Yes, I've done this for now just to fix the build. This is part of the list view bulk actions which we don't currently have in V5. I think it can be fixed when we work on that

Marc-Roig commented on packages/core/content-manager/server/src/controllers/collection-types.ts:
```diff
@@ -411,10 +413,12 @@ export default {
     const { model } = ctx.params;
     const { body } = ctx.request;
-    const { ids } = body;
+    const { documentIds } = body;
```
cc @Feranchz

Marc-Roig commented on packages/core/content-manager/server/src/controllers/collection-types.ts:
```diff
@@ -429,8 +433,10 @@ export default {
 
-    const entityPromises = ids.map((id: any) => documentManager.findOne(id, model, { populate }));
-    const entities = await Promise.all(entityPromises);
+    const entityPromises = documentIds.map((documentId: any) =>
+      documentManager.findLocales(documentId, model, { populate, locale })
+    );
+    const entities = (await Promise.all(entityPromises)).flat();
```
Was thinking about creating an `async.flatMap`  utility for cases like this, wdyt? do you think its worth in cases like this?

Marc-Roig commented on packages/core/content-manager/server/src/services/document-manager.ts:
```diff
@@ -193,44 +193,27 @@ const documentManager = ({ strapi }: { strapi: Core.Strapi }) => {
-          );
-        })
-      );
+    async publishMany(uid: UID.ContentType, documentIds: string[], locale?: string | string[]) {
+      let publishedEntitiesCount = 0;
+      await async.map(documentIds, async (documentId: string) => {
```
this work might collide with @Feranchz , he is working on bulk publsihing

Feranchz replied:
Yes, but this PR is more advanced than mine and in any case I was doing something **very** similar so maybe we can go ahead with this. 

I will change my PR's scope to only the bulkDelete which is not changed in this PR and follow this

Marc-Roig commented on packages/core/types/src/modules/documents/params/document-engine.ts:
```diff
@@ -69,7 +77,7 @@ export type Publish<TContentTypeUID extends UID.ContentType> = Pick<
 export type Unpublish<TContentTypeUID extends UID.ContentType> = Pick<
   TContentTypeUID,
-  'filters' | 'locale' | 'lookup'
+  'filters' | 'locale:string' | 'lookup'
```
do we plan to support unpublishing many locales too? I find it limiting that you can publish many of them but not unpublish.

jhoward1994 replied:
Yep we do, we're just handling publishing first. Once everything is in place for
publishing it shouldn't be too much work to handle unpublishing

Marc-Roig commented on packages/core/core/src/services/document-service/internationalization.ts:
```diff
@@ -69,7 +54,7 @@ const localeToData: Transform = (contentType, params) => {
   }
 
-  if (params.locale) {
+  if (params.locale && typeof params.locale === 'string' && params.locale !== '*') {
     return assoc(['data', 'locale'], params.locale, params);
   }
```
should we throw an error if users try to use a locale that's invalid?

What happens if a user is creating a document but sends a locale that's not a string?

Marc-Roig commented on packages/core/core/src/services/document-service/internationalization.ts:
```diff
@@ -36,22 +36,7 @@ const localeToLookup: Transform = (contentType, params) => {
-
-  if (params.locale) {
-    if (params.locale === '*') {
+    if (typeof params.locale === 'string' && params.locale === '*') {
```
won't this change allow doing a find many on multiple locales at the same time?
If we want to prevent that I believe we should keep the localetoLookup & multiLocaletoLookup

Feranchz commented on packages/core/content-manager/server/src/services/document-manager.ts:
```diff
@@ -193,44 +193,27 @@ const documentManager = ({ strapi }: { strapi: Core.Strapi }) => {
+      let publishedEntitiesCount = 0;
+      await async.map(documentIds, async (documentId: string) => {
+        // Publish per documentId
+        const res = await this.publish(documentId, uid, { locale });
+        publishedEntitiesCount += res?.length ?? 0;
```
What happens if one publish failed? I think we throw an error so we can end with only the half of the documentIds published even if we throw an error.

I think we should wrapper this in a transaction.

I think we can save the `publishedEntitiesCount`, because we publish all or we don't publish anything, so at the end we only need to return `documentIds.length`

Feranchz commented on packages/core/content-manager/shared/contracts/collection-types.ts:
```diff
@@ -305,15 +305,21 @@ export declare namespace BulkPublish {
       documentIds: Modules.Documents.ID[];
     };
-    query: {};
+    query: {
+      // If not provided, the default locale will be used
+      // Otherwise specify the locales to publish of the documents
+      locale?: string | string[] | null;
+    };
```
Do we have some idea of how we are going to use this BulkPublish for multiple locales? 🤔 Because I was thinking in a API like:
```
body:

[ { documentId: 'abcd', locales: ['en'] }, { documentId: 'qwerty', locales: ['en', 'es'] } ]
```

So, in one action we can publish multiple entities in the locales we are interested for that specific entity

cc @Marc-Roig

jhoward1994 replied:
Do we have this use case in the CM? 

I was thinking that edit view bulk publish = single locale across multiple
documents whereas list view bulk publish (this PR) is multiple locales of the
same document.

Regardless I think passing multiple locales and documentIds should work in a way
like:

"publish all of these locales of all of these documents if they exist"

e.g. for documents that don't have some of the locales passed they'd just be
ignored

What do you think?

Feranchz replied:
I agree, I wasn't considering the new bulkPublishs options we have in the editView

jhoward1994 commented on packages/core/core/src/services/document-service/internationalization.ts:
```diff
@@ -36,7 +37,15 @@ const localeToLookup: Transform = (contentType, params) => {
+      return assoc(['lookup', 'locale'], params.locale, params);
+    }
+
+    throw new errors.ValidationError(
```
Revealed a bug being worked on https://github.com/strapi/strapi/pull/20185

innerdvations commented on packages/core/content-manager/server/src/services/document-manager.ts:
```diff
@@ -193,53 +193,18 @@ const documentManager = ({ strapi }: { strapi: Core.Strapi }) => {
-      const publishedEntitiesCount = await strapi.db.query(uid).updateMany({
-        where: filters,
-        data,
+        const publishedEntitiesCount = results.flat().filter(Boolean).length;
```
Won't this actually return the number of documents that were published, not entities?

jhoward1994 replied:
I don't think so, each call to publish (per document id) is an array of the published
entities for that documentId

So if we flatten that array it is the total number of entities?

innerdvations replied:
ahhh, ok, I missed that, sorry!

innerdvations commented on packages/core/content-manager/server/src/services/document-metadata.ts:
```diff
@@ -85,26 +95,56 @@ export default ({ strapi }: { strapi: Core.Strapi }) => ({
+    const mappingResult = await async.map(
+      Object.values(versionsByLocale),
+      async (localeVersions: DocumentVersion[]) => {
+        const model = strapi.getModel(uid);
```
This could be defined just once outside this function

innerdvations commented on packages/core/content-manager/server/src/services/document-metadata.ts:
```diff
@@ -85,26 +95,56 @@ export default ({ strapi }: { strapi: Core.Strapi }) => ({
+
+        const mappedLocaleVersions: DocumentVersion[] = await async.map(
+          localeVersions,
+          async (localeVersion: DocumentVersion) =>
```
It looks like this method could also be declared just once, outside all of these nested maps unless I'm missing something which is very possible 😆

jhoward1994 replied:
thanks!

innerdvations commented on packages/core/core/src/services/document-service/internationalization.ts:
```diff
@@ -36,7 +37,15 @@ const localeToLookup: Transform = (contentType, params) => {
   }
 
   if (params.locale) {
-    return assoc(['lookup', 'locale'], params.locale, params);
+    const isValidLocale = typeof params.locale === 'string';
+
+    if (isValidLocale) {
+      return assoc(['lookup', 'locale'], params.locale, params);
+    }
+
+    throw new errors.ValidationError(
+      `Invalid locale param ${params.locale} provided. Document locales must be strings.`
+    );
   }
 
   return params;
```
Primary suggestion is `String(params.locale)` (or maybe there's something safer/better) since we're trying to display as string something we explicitly know isn't a string.

But could we rewrite it more like this to make it less nested?

```
const localeToLookup: Transform = (contentType, params) => {
  if (
    !params.locale ||
    !strapi.plugin('i18n').service('content-types').isLocalizedContentType(contentType)
  ) {
    return params;
  }

  if (typeof params.locale !== 'string') {
    throw new errors.ValidationError(
      `Invalid locale param ${String(params.locale)} provided. Document locales must be strings.`
    );
  }

  return assoc(['lookup', 'locale'], params.locale, params);
};
```

innerdvations replied:
And maybe a comment about why `'*'` is allowed here when it's not in the others?

innerdvations commented on packages/core/utils/src/__tests__/content-types.test.ts:
```diff
@@ -110,6 +111,30 @@ describe('Content types utils', () => {
+    test('false for field without constraints', () => {
+      expect(getDoesAttributeRequireValidation({ type: 'string' })).toEqual(false);
+    });
+    test('true for required fields', () => {
+      expect(getDoesAttributeRequireValidation({ type: 'string', required: true })).toEqual(true);
+    });
+    test('true for unique fields', () => {
+      expect(getDoesAttributeRequireValidation({ type: 'string', unique: true })).toEqual(true);
+    });
+    test('true for numeric fields with a max', () => {
+      expect(getDoesAttributeRequireValidation({ type: 'integer', max: 10 })).toEqual(true);
+    });
+    test('true for numeric fields with a min', () => {
+      expect(getDoesAttributeRequireValidation({ type: 'integer', min: 10 })).toEqual(true);
+    });
+    test('true for fields with a max length', () => {
+      expect(getDoesAttributeRequireValidation({ type: 'string', maxLength: 10 })).toEqual(true);
+    });
+    test('true for fields with a min length', () => {
+      expect(getDoesAttributeRequireValidation({ type: 'string', minLength: 10 })).toEqual(true);
+    });
```
Maybe also test the falsy versions of each of these should return false (for required and unique at least)

jhoward1994 replied:
Done here [`e2c5b0c` (#20046)](https://github.com/strapi/strapi/pull/20046/commits/e2c5b0c6a6a04d0d6a2a685057798d85c89bde96)

innerdvations commented on tests/api/core/content-manager/api/document-metadata.test.api.js:
```diff
@@ -174,4 +234,140 @@ describe('CM API - Document metadata', () => {
+    expect(meta.availableStatus).toEqual([]);
+  });
+
+  test('Does not return any sensitive information in the available status', async () => {
```
I wonder if either this or the previous test could use `toEqual` / `toContainEqual` to ensure we don't get any unexpected data? I know that kind of test is more of a pain to write and keep updated as things change, but in this case it might be better

innerdvations commented on tests/api/core/content-manager/api/number-of-draft-relations.test.api.js:
```diff
@@ -397,13 +430,65 @@ describe('CM API - Number of draft relations', () => {
+
+    // 4 (newly created default locale) + 8 (non default locale of the same
+    // document) + 12 (previously created default locale) = 24
+    expect(body.data).toBe(24);
```
Since this depends on previous tests running, you can't use `.only` on this test and if anyone adds a test they will have to update this. Could it do a db reset, create some data, then count that?

innerdvations reviewed (commented):
Looking good, just a few comments

jhoward1994 commented on packages/core/content-manager/server/src/history/services/history.ts:
```diff
@@ -168,6 +168,14 @@ const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
         const defaultLocale = await getDefaultLocale();
+
+        if (Array.isArray(documentContext?.locale)) {
+          // TODO calls picked from the middleware could contain an array of
```
FYI @remidej @markkaylor

jhoward1994 commented on packages/core/content-manager/server/src/controllers/collection-types.ts:
```diff
@@ -394,27 +393,31 @@ export default {
         // data: setCreatorFields({ user, isEdition: true })({}),
       });
+
+      return publishResult.at(0);
```
Done in [`e2c5b0c` (#20046)](https://github.com/strapi/strapi/pull/20046/commits/e2c5b0c6a6a04d0d6a2a685057798d85c89bde96)

jhoward1994 commented on packages/core/content-manager/server/src/controllers/validation/__tests__/dimensions.test.ts:
```diff
@@ -0,0 +1,30 @@
+    ).rejects.toThrow();
+  });
+
+  test('neither status or locale are required', () => {
```
[`e2c5b0c` (#20046)](https://github.com/strapi/strapi/pull/20046/commits/e2c5b0c6a6a04d0d6a2a685057798d85c89bde96)

jhoward1994 commented:
@innerdvations thanks for the feedback. I was just commenting the same link on all your comments. I dont know why sorry 😅

Anyway, I've addressed your feedback in [`e2c5b0c` (#20046)](https://github.com/strapi/strapi/pull/20046/commits/e2c5b0c6a6a04d0d6a2a685057798d85c89bde96)

innerdvations reviewed (approved):
Nice! LGTM, I will QA the front-end side of it

jhoward1994 commented:
Thanks! I'll merge this all into the feature `[feat/bulk-locale-publish](https://github.com/strapi/strapi/tree/feat/bulk-locale-publish)` for QA
